streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java

```diff
@@ -462,6 +583,12 @@ private void cleanRemovedTasksCalledByCleanerThread(final long cleanupDelayMs) {
                         final long now = time.milliseconds();
                         final long lastModifiedMs = taskDir.file().lastModified();
                         if (now - cleanupDelayMs > lastModifiedMs) {
+                            final Task task = tasksForLocalState.remove(id);
+                            if (task != null) {
+                                // close the pending Task, if one still exists
+                                task.suspend();
+                                task.closeClean();
+                            }
```

Comment:
Wouldn't the existence of a task be an illegal state?

Reply:
It's technically possible to reach this is the cleaner thread runs _before_ the first assignment completes for all threads. `tasksForLocalState` could, in this case, contain some tasks that could be stale enough to need to be cleaned.

Reply:
I am not sure I understand. The cleaner thread only deletes the task directory if it is not locked. However, the task directory is locked for a pending task until it is assigned (lock is replaced) or it is closed because it is not assigned (unlocked after it is removed from the internal map `tasksForLocalState`). Where in this lifecycle can the cleaner thread find a pending task that is not locked and is still in the internal map `tasksForLocalState`?

Reply:
Good point! I hadn't considered that. I've removed this in a07c268